[Dwarf] Support `__ptrauth` qualifier in metadata nodes

kovdan01 opened the pull request:
Reland #82363 after fixing build failure https://lab.llvm.org/buildbot/#/builders/5/builds/41428.

Memory sanitizer detects usage of `RawData` union member which is not
filled directly. Instead, the code relies on filling `Data` union
member, which is a struct consisting of signing schema parameters.

According to https://en.cppreference.com/w/cpp/language/union, this is UB:
"It is undefined behavior to read from the member of the union that
wasn't most recently written".

Instead of relying on compiler allowing us to do dirty things, do not
use union and only store `RawData`. Particular ptrauth parameters are
obtained on demand via bit operations.

Original PR description below.

Emit `__ptrauth`-qualified types as `DIDerivedType` metadata nodes in IR with tag `DW_TAG_LLVM_ptrauth_type`, baseType referring to the type which has the qualifier applied, and the following parameters representing the signing schema:

- `ptrAuthKey` (integer)
- `ptrAuthIsAddressDiscriminated` (boolean)
- `ptrAuthExtraDiscriminator` (integer)
- `ptrAuthIsaPointer` (boolean)
- `ptrAuthAuthenticatesNullValues` (boolean)

Co-authored-by: Ahmed Bougacha <ahmed@bougacha.org>

kovdan01 committed:
[Dwarf] Support `__ptrauth` qualifier in metadata nodes

Emit `__ptrauth`-qualified types as `DIDerivedType` metadata nodes in IR with
tag `DW_TAG_LLVM_ptrauth_type`, baseType referring to the type which has the
qualifier applied, and the following parameters representing the signing
schema:

- `ptrAuthKey` (integer)
- `ptrAuthIsAddressDiscriminated` (boolean)
- `ptrAuthExtraDiscriminator` (integer)
- `ptrAuthIsaPointer` (boolean)
- `ptrAuthAuthenticatesNullValues` (boolean)

kovdan01 committed:
Merge branch 'main' into dwarf-pauth-llvm-dinodes

kovdan01 committed:
Restore accidentally removed line

kovdan01 committed:
Do not use else after return

kovdan01 committed:
Do not include llvm/BinaryFormat/Dwarf.h in llvm/IR/DebugInfoMetadata.h

kovdan01 committed:
[MLIR] Fix `llvm::DIDerivedType::get` call (new argument PtrAuthData proposed)

kovdan01 committed:
Fix incorrect union member usage

See build failure
https://lab.llvm.org/buildbot/#/builders/5/builds/41428.

Memory sanitizer detects usage of `RawData` union member which is not
filled directly. Instead, the code relies on filling `Data` union
member, which is a struct consisting of signing schema parameters.

According to https://en.cppreference.com/w/cpp/language/union, this is UB:
"It is undefined behavior to read from the member of the union that
wasn't most recently written".

Instead of relying on compiler allowing us to do dirty things, do not
use union and only store `RawData`. Particular ptrauth parameters are
obtained on demand via bit operations.

asl commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -975,34 +975,29 @@ class DIDerivedType : public DIType {
-      Payload.Data.ExtraDiscriminator = Discriminator;
-      Payload.Data.IsaPointer = IsaPointer;
-      Payload.Data.AuthenticatesNullValues = AuthenticatesNullValues;
+      RawData = (Key << 0) | (IsDiscr ? (1 << 4) : 0) | (Discriminator << 5) |
```
Please add comment describing the layout of `RawData`.

kovdan01 replied:
Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

dwblaikie commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
+                (IsaPointer ? (1 << 21) : 0) |
+                (AuthenticatesNullValues ? (1 << 22) : 0);
+    }
+    bool operator==(struct PtrAuthData Other) const {
+      return RawData == Other.RawData;
+    }
+    bool operator!=(struct PtrAuthData Other) const {
+      return !(*this == Other);
+    }
+
```
Prefer non-member operator overloads (they can be friends and defined inline if that's useful/more convenient than defining them outside the class definition) to allow equal conversions on LHS and RHS.

kovdan01 replied:
Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

dwblaikie commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
+                (IsaPointer ? (1 << 21) : 0) |
+                (AuthenticatesNullValues ? (1 << 22) : 0);
+    }
+    bool operator==(struct PtrAuthData Other) const {
```
Drop the `struct` qualifier as it's not needed (well, it's needed if the name is overloaded to be a struct and something else, but probably best not to do that) in C++.

kovdan01 replied:
Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

dwblaikie commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -1048,6 +1088,39 @@ class DIDerivedType : public DIType {
+  std::optional<PtrAuthData> getPtrAuthData() const;
+
+  /// \returns The PointerAuth key.
+  std::optional<unsigned> getPtrAuthKey() const {
```
Are these helpers particularly useful? It seems like they change the usage code from this:
```
if (auto PtrAuthData = getPtrAuthData())
  do_something_with(PtrAuthData->Key());
```
to
```
if (auto Key = getPtrAuthKey())
  do_something_with(*Key);
```

Which doesn't seem like a huge win?

kovdan01 replied:
These are just small shortcuts which do not give any value except what you described above. I personally have no strong preference, so accepting your suggestion - deleted the helpers. See f765247a3d49b3b726a7442d8d31c1cbd4782154

kovdan01 committed:
Merge branch 'main' into dwarf-pauth-llvm-dinodes

kovdan01 committed:
Address review comments

kovdan01 committed:
Merge branch 'main' into dwarf-pauth-llvm-dinodes

kovdan01 commented:
Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes

MaskRay commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
+    }
+
+    unsigned Key() { return (RawData >> 0) & 0b1111; }
+    bool IsAddressDiscriminated() { return (RawData >> 4) & 1; }
```
Nit: `functionName`. The code base is unfortunately inconsistent, but newer functions mostly stick with the recommended convention.

kovdan01 replied:
Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117

MaskRay commented on llvm/lib/AsmParser/LLParser.cpp:
```diff
@@ -5145,19 +5149,30 @@ bool LLParser::parseDIDerivedType(MDNode *&Result, bool IsDistinct) {
     DWARFAddressSpace = dwarfAddressSpace.Val;
+  std::optional<DIDerivedType::PtrAuthData> PtrAuthData;
+  if (ptrAuthKey.Val)
+    PtrAuthData = DIDerivedType::PtrAuthData(
```
Nit: `emplace` https://en.cppreference.com/w/cpp/utility/optional/emplace

kovdan01 replied:
Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117

MaskRay commented on llvm/test/DebugInfo/AArch64/ptrauth.ll:
```diff
@@ -0,0 +1,70 @@
+
+target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
+
+@p = common global i8* null, align 8, !dbg !0
```
`i8*` indicates a typed pointer, which LLVM is migrating away. Use `ptr`.
Consider a non-common linkage. COMMON linkage is also obsoleted.

kovdan01 replied:
Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117. Regarding linkage - any type is applicable for the test, so just removed that, and by default that should be `external`.

kovdan01 committed:
Address review comments

kovdan01 commented on llvm/lib/IR/DIBuilder.cpp:
```diff
@@ -296,7 +296,20 @@ DIStringType *DIBuilder::createStringType(StringRef Name,
+  return DIDerivedType::get(VMContext, dwarf::DW_TAG_LLVM_ptrauth_type, "",
+                            nullptr, 0, nullptr, FromTy, 0, 0, 0, std::nullopt,
+                            std::optional<DIDerivedType::PtrAuthData>(
+                                std::in_place, Key, IsAddressDiscriminated,
```
According to https://llvm.org/docs/CodingStandards.html#c-standard-versions, we can rely on C++17, so use `std::in_place` for in-place initialization of `std::optional` (see (6) https://en.cppreference.com/w/cpp/utility/optional/optional). Please let me know if we for some reason can't rely on this C++17 feature here.

kovdan01 commented:
Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes.

@MaskRay I've addressed your comments in 1adcade6f8d54d8e8ebebc52709e4b293450d117 - would be glad if you let me know if the fixes are OK.

JDevlieghere reviewed (approved):
I would've probably done the inverse and shifted the `RawData` in the constructor, stored them as individual fields in a struct and have the getters operate on that. And of course have a new vector to return the raw data. But it looks like we rely on `RawData` more than on the getters, so this might be slightly more efficient. LGTM!

adrian-prantl commented on llvm/lib/Bitcode/Reader/MetadataLoader.cpp:
```diff
@@ -1566,8 +1566,17 @@ Error MetadataLoader::MetadataLoaderImpl::parseOneMetadata(
+    // If not, we can't tell which was intended to be embedded, as both ptrauth
+    // and annotations have been expected at Record[13] at various times.
+    if (Record.size() > 14) {
+      if (Record[13])
```
This if  is redundant?

MaskRay commented:
> Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes.
> 
> @MaskRay I've addressed your comments in [1adcade](https://github.com/llvm/llvm-project/commit/1adcade6f8d54d8e8ebebc52709e4b293450d117) - would be glad if you let me know if the fixes are OK.

Thanks. I'm trying to study this patch as I am not so familiar with debug info IR.

If you don't mind waiting a bit, waiting for @dwblaikie (another debug information code owner) could be useful.

MaskRay commented:
I was confused as I haven't attended any [LLVM Pointer Authentication sync-ups](https://discourse.llvm.org/t/llvm-pointer-authentication-sync-ups/62661) and I am definitely lacking context. It seems that there is an upstreaming proposal back in 2019 https://discourse.llvm.org/t/rfc-pointer-authentication-for-arm64e/53433 and apple/llvm-project contains these `DW_TAG_LLVM_ptrauth_type` changes. This PR is part of the upstream efforts. Yeah, looks good to me.

Normally, when I see `createPtrAuthQualifiedType` with no in-tree user, I'd ask that you prepare links for the immediate user.

JDevlieghere commented:
@dwblaikie Did you want to have a look at this before we merge this? This PR is going to cause quite a bit of conflicts downstream so I would be nice to merge this somewhat early in the week.

dwblaikie commented:
> @dwblaikie Did you want to have a look at this before we merge this? This PR is going to cause quite a bit of conflicts downstream so I would be nice to merge this somewhat early in the week.

*glances* What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?

JDevlieghere commented:
> What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?

Just regular merge conflicts when we merge this into the Swift fork. This PR is upstreaming ptrauth support and started off based on the downstream code, but after a few iterations of review it looks a bit different. That's totally fine, as the result is better for it, but git will need a hand with the merge. The only reason I'm bringing it up is because last time this PR landed on a Friday afternoon (before it got revert) which left us little to deal with all this.

dwblaikie commented:
> > What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?
> 
> Just regular merge conflicts when we merge this into the Swift fork. This PR is upstreaming ptrauth support and started off based on the downstream code, but after a few iterations of review it looks a bit different. That's totally fine, as the result is better for it, but git will need a hand with the merge. The only reason I'm bringing it up is because last time this PR landed on a Friday afternoon (before it got revert) which left us little to deal with all this.

Ah, your own downstream, fair enough. It looks right enough to me, FWIW :)
